### PR TITLE
source: add an upper bound timeout to HTTP requests

### PR DESCRIPTION
**Pull request by lucab**
This adds a very large timeout (4 hours) to all HTTP requests, as they
currently default to none.
It is meant to almost never trigger, except under pathological network
conditions. In such cases, it should help detecting issues by explicitly
erroring out instead of hanging forever.

### CODE REVIEW

**Comment by cgwalters** [comment]
Does this apply even in the face of retrying requests?  (Do we retry requests?)

**Comment by lucab** [comment]
This is a completion timeout on each request. In case of N retries, the overall waiting time would be `N * timeout` in the worst case.

Additionally, I think we aren't currently retrying HTTP downloads. I guess this is in part because we directly stream the response/image to disk. I'm unsure whether there should be a top-level retry logic with a `--retry` flag, or if we should simply leave that to a looping caller.

/cc @bgilbert

File: src/source.rs [hunk 1]
```diff
@@ -20,10 +20,15 @@ use std::fmt::{Display, Formatter};
 use std::fs::OpenOptions;
 use std::io::{Read, Seek, SeekFrom};
 use std::path::Path;
+use std::time::Duration;
 
 use crate::cmdline::*;
 use crate::errors::*;
 
+/// Default timeout for HTTP requests (4 hours).
+const DEFAULT_HTTP_REQ_TIMEOUT: Duration = Duration::from_secs(4 * 60 * 60);
```
Comment (bgilbert) [inline src/source.rs]: Let's not call this the "default" since it's not configurable.

Comment (lucab) [reply src/source.rs]: Ack, I've renamed to `HTTP_COMPLETION_TIMEOUT`. Were you hinting at keeping a ticket around to maybe later add a CLI flag to configure this, or do you think it isn't useful to have that?

Comment (bgilbert) [reply src/source.rs]: Nah, I don't think it'd be useful.  We can always add later if needed.

**Pull request merged**